dao-contracts/src/kyc_voter.rs

```diff
@@ -1,17 +1,144 @@
+    casper_dao_macros::{casper_contract_interface, Instance},
+    casper_env::{self, caller},
+    Address, Error, SequenceGenerator,
+};
+use casper_types::{runtime_args, RuntimeArgs, U256};
+
+use crate::voting::{kyc::KycInfo, voting::Voting, Ballot, Choice, GovernanceVoting, VotingId};
+use delegate::delegate;
+
+const ARG_TO: &str = "to";
+const ARG_TOKEN_ID: &str = "token_id";
+const ENTRY_POINT_MINT: &str = "mint";
+
+#[casper_contract_interface]
+pub trait KycVoterContractInterface {
+    fn init(&mut self, variable_repo: Address, reputation_token: Address, kyc_token: Address);
+    // Require no voting for a given `address` is on.
+    // Precondition: KycNft.balance_of(address_to_onboard) == 0;
+    // Action: KycNft.mint(address_to_onboard, next_token_id)
+    fn create_voting(&mut self, address_to_onboard: Address, document_hash: String, stake: U256);
+    fn vote(&mut self, voting_id: VotingId, choice: Choice, stake: U256);
+    fn finish_voting(&mut self, voting_id: VotingId);
+    fn get_dust_amount(&self) -> U256;
+    fn get_variable_repo_address(&self) -> Address;
+    fn get_reputation_token_address(&self) -> Address;
+    fn get_kyc_token_address(&self) -> Address;
+    fn get_voting(&self, voting_id: VotingId) -> Option<Voting>;
+    fn get_ballot(&self, voting_id: VotingId, address: Address) -> Option<Ballot>;
+    fn get_voter(&self, voting_id: VotingId, at: u32) -> Option<Address>;
+    fn get_document_hash(&self, voting_id: VotingId) -> Option<String>;
+}
+
+#[derive(Instance)]
+pub struct KycVoterContract {
+    kyc: KycInfo,
+    voting: GovernanceVoting,
+    sequence: SequenceGenerator,
+}
+
+impl KycVoterContractInterface for KycVoterContract {
+    fn init(&mut self, variable_repo: Address, reputation_token: Address, kyc_token: Address) {
+        self.kyc.init(kyc_token);
+        self.voting.init(variable_repo, reputation_token);
+    }
+
+    delegate! {
+        to self.kyc {
+            fn get_kyc_token_address(&self) -> Address;
+            fn get_document_hash(&self, voting_id: VotingId) -> Option<String>;
+        }
+
+        to self.voting {
+            fn get_variable_repo_address(&self) -> Address;
+            fn get_reputation_token_address(&self) -> Address;
+            fn get_dust_amount(&self) -> U256;
+            fn get_voting(&self, voting_id: U256) -> Option<Voting>;
+            fn get_ballot(&self, voting_id: U256, address: Address) -> Option<Ballot>;
+            fn get_voter(&self, voting_id: U256, at: u32) -> Option<Address>;
+        }
+    }
+
+    fn create_voting(&mut self, address_to_onboard: Address, document_hash: String, stake: U256) {
+        self.assert_no_ongoing_voting(&address_to_onboard);
+        self.assert_not_kyced(&address_to_onboard);
+
+        let creator = caller();
+        let contract_to_call = self.get_kyc_token_address();
+        let token_id = self.sequence.next_value();
+        let runtime_args = runtime_args! {
```

Comment:
You can get rid of the whole `documents` storage if you use `document_hash` as a `ARG_TOKEN_ID`. And this what actually should happen, so the KYC NFT Contract itself holds the link to the KYC document.
Sorry for not specifying that more explicitly earlier.